Sources/Datadog/Tracing/Utils/MethodSwizzler.swift

```diff
@@ -0,0 +1,77 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal enum SwizzlingError: LocalizedError {
+    case unknown
+    case classIsNotNSObjectSubclass(className: String)
+    case methodNotFound(selector: String, className: String)
+    case methodWasNotSwizzled(selector: String, className: String)
+}
+
+internal class MethodSwizzler {
+    static let shared = MethodSwizzler()
+
+    private let queue = DispatchQueue(label: "com.datadoghq.methodSwizzlerQueue", target: DispatchQueue.global(qos: .userInteractive))
+    private var implementationCache = [String: IMP]()
+
+    func currentImplementation<TypedIMP>(of selector: Selector, in klass: AnyClass) throws -> TypedIMP {
+        try queue.sync {
+            let foundMethod = try method(with: selector, in: klass)
+            return unsafeBitCast(method_getImplementation(foundMethod), to: TypedIMP.self)
+        }
+    }
+
+    func swizzle(selector: Selector, in klass: AnyClass, with implementation: IMP) throws {
+        try queue.sync {
+            if !class_respondsToSelector(klass, #selector(NSObject.perform(_:))) {
+                throw SwizzlingError.classIsNotNSObjectSubclass(className: NSStringFromClass(klass))
+            }
+            let foundMethod = try method(with: selector, in: klass)
+            let cacheKey = methodIdentifier(foundMethod, in: klass)
+            if implementationCache[cacheKey] == nil {
+                implementationCache[cacheKey] = method_getImplementation(foundMethod)
+            }
+            method_setImplementation(foundMethod, implementation)
+        }
+    }
+
+    func unswizzle(selector: Selector, in klass: AnyClass) throws {
```

Comment:
We don't need the unswizzling in production code. Let's make it an `internal extension` in `DatadogTests` target. We'll move it to `Datadog` once needed

Reply:
then we would need to expose `implementationCache` which is worse.
this keeps everything in one place

Reply:
OK I see, the `implementationCache` is necessary for unswizzling which we need in tests. Although I don't like it - it runs the code which we need only for tests, I don't see any good way to abstract it just for tests.